README.md
# Logstash Plugin

This is a plugin for [Logstash](https://github.com/elastic/logstash). It support [SolrCloud](https://cwiki.apache.org/confluence/display/solr/SolrCloud) not only Standalone Solr.

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Config parameters

### url

The Solr server url (for example http://localhost:8983/solr/collection1).

```
url http://localhost:8983/solr/collection1
```

### zk_host

The ZooKeeper connection string that SolrCloud refers to (for example localhost:2181/solr).

```
zk_host localhost:2181/solr
```

### collection

The SolrCloud collection name (default collection1).

```
collection collection1
```

### defined_fields

The defined fields in the Solr schema.xml. If omitted, it will get fields via Solr Schema API.

```
defined_fields ["id", "title"]
```

### ignore_undefined_fields

Ignore undefined fields in the Solr schema.xml.

```
ignore_undefined_fields false
```

### unique_key_field

A field name of unique key in the Solr schema.xml. If omitted, it will get unique key via Solr Schema API.

```
unique_key_field id
```

### timestamp_field

A field name of event timestamp in the Solr schema.xml (default event_timestamp).

```
timestamp_field event_timestamp
```

### flush_size

A number of events to queue up before writing to Solr (default 100).

```
flush_size 100
```

## Plugin setup examples

### Sent to standalone Solr using data-driven schemaless mode.
```
output {
  solr {
    url => "http://localhost:8983/solr/collection1"
  }
}

```

### Sent to SolrCloud using data-driven schemaless mode.
```
output {
  solr {
    zk_host => "localhost:2181/solr"
    collection => "collection1"
  }
}
```

## Solr setup examples

### How to setup Standalone Solr using data-driven schemaless mode.

1.Download and install Solr

```sh
$ mkdir $HOME/solr
$ cd $HOME/solr
$ wget https://archive.apache.org/dist/lucene/solr/5.4.0/solr-5.4.0.tgz
$ tar zxvf solr-5.4.0.tgz
$ cd solr-5.4.0
```

2.Start standalone Solr

```sh
$ ./bin/solr start -p 8983 -s server/solr
```

3.Create core

```sh
$ ./bin/solr create -c collection1 -d server/solr/configsets/data_driven_schema_configs -n collection1_configs
```

### How to setup SolrCloud using data-driven schemaless mode (shards=1 and replicationfactor=2).

1.Download and install ZooKeeper

```sh
$ mkdir $HOME/zookeeper
$ cd $HOME/zookeeper
$ wget https://archive.apache.org/dist/zookeeper/zookeeper-3.4.6/zookeeper-3.4.6.tar.gz
$ tar zxvf zookeeper-3.4.6.tar.gz
$ cd zookeeper-3.4.6
$ cp -p ./conf/zoo_sample.cfg ./conf/zoo.cfg
```

2.Start standalone ZooKeeper

```sh
$ ./bin/zkServer.sh start
```

3.Download an install Solr

```sh
$ mkdir $HOME/solr
$ cd $HOME/solr
$ wget https://archive.apache.org/dist/lucene/solr/5.4.0/solr-5.4.0.tgz
$ tar zxvf solr-5.4.0.tgz
$ cd solr-5.4.0
$ ./server/scripts/cloud-scripts/zkcli.sh -zkhost localhost:2181 -cmd clear /solr
$ ./server/scripts/cloud-scripts/zkcli.sh -zkhost localhost:2181 -cmd makepath /solr
$ cp -pr server/solr server/solr1
$ cp -pr server/solr server/solr2
```

4.Start SolrCloud

```sh
$ ./bin/solr start -h localhost -p 8983 -z localhost:2181/solr -s server/solr1
$ ./bin/solr start -h localhost -p 8985 -z localhost:2181/solr -s server/solr2
```

5.Create collection

```sh
$ ./bin/solr create -c collection1 -d server/solr1/configsets/data_driven_schema_configs -n collection1_configs -shards 1 -replicationFactor 2
```

## Need Help?

Need help? Try #logstash on freenode IRC or the [https://discuss.elastic.co/c/logstash](https://discuss.elastic.co/c/logstash) discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:

```ruby
gem "logstash-output-solr", :path => "/your/local/logstash-output-solr"
```

- Install plugin

```sh
bin/plugin install --no-verify
```

- Run Logstash with your plugin

```sh
bin/logstash -e 'output {solr {}}'
```

At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem

```sh
gem build logstash-output-solr.gemspec
```

- Install the plugin from the Logstash home

```sh
bin/plugin install /your/local/plugin/logstash-output-solr.gem
```

- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.

CHANGELOG.md
## 0.3.0
 - Update rsolr-cloud.

## 0.2.0
 - Bug fix.

## 0.2.0
 - Supports SolrCloud.

## 0.1.0
 - First release.


spec/outputs/solr_spec.rb
# encoding: utf-8
require "logstash/devutils/rspec/spec_helper"
require "logstash/outputs/solr"
require "logstash/codecs/plain"
require "logstash/event"

require 'zk-server'
require 'zk'

describe LogStash::Outputs::Solr do

  before do
    @zk_server = nil
  end

  describe 'configuration' do
    let(:config) {
      {
        'url' => 'http://localhost:8983/solr/collection1',
        'zk_host' => 'localhost:2181/solr',
        'collection' => 'collection1',
        'defined_fields' => ['id', 'title'],
        'ignore_undefined_fields' => true,
        'unique_key_field' => 'id',
        'timestamp_field' => 'event_timestamp',
        'flush_size' => 100
      }
    }

    it 'url' do
      output = LogStash::Outputs::Solr.new(config)
      expect(output.config['url']).to eq('http://localhost:8983/solr/collection1')
    end

    it 'zk_host' do
      output = LogStash::Outputs::Solr.new(config)
      expect(output.config['zk_host']).to eq('localhost:2181/solr')
    end

    it 'collection' do
      output = LogStash::Outputs::Solr.new(config)
      expect(output.config['collection']).to eq('collection1')
    end

    it 'defined_fields' do
      output = LogStash::Outputs::Solr.new(config)
      expect(output.config['defined_fields']).to eq(['id', 'title'])
    end

    it 'ignore_undefined_fields' do
      output = LogStash::Outputs::Solr.new(config)
      expect(output.config['ignore_undefined_fields']).to eq(true)
    end
    
    it 'unique_key_field' do
      output = LogStash::Outputs::Solr.new(config)
      expect(output.config['unique_key_field']).to eq('id')
    end

    it 'timestamp_field' do
      output = LogStash::Outputs::Solr.new(config)
      expect(output.config['timestamp_field']).to eq('event_timestamp')
    end

    it 'flush_size' do
      output = LogStash::Outputs::Solr.new(config)
      expect(output.config['flush_size']).to eq(100)
    end
  end

  describe 'register_standalone' do
    let(:config) {
      {
        'url' => 'http://localhost:8983/solr/collection1',
        'defined_fields' => ['id', 'title'],
        'ignore_undefined_fields' => true,
        'unique_key_field' => 'id',
        'timestamp_field' => 'event_timestamp',
        'flush_size' => 100
      }
    }

    it 'mode' do
      output = LogStash::Outputs::Solr.new(config)
      output.register

      mode = output.instance_variable_get('@mode')
      expect(mode).to eq('Standalone')
    end
  end

  describe 'register_solrcloud' do
    before do
      start_zookeeper
    end

    after do
      stop_zookeeper
    end

    let(:config) {
      {
        'zk_host' => 'localhost:3292/solr',
        'collection' => 'collection1',
        'defined_fields' => ['id', 'title'],
        'ignore_undefined_fields' => true,
        'unique_key_field' => 'id',
        'timestamp_field' => 'event_timestamp',
        'flush_size' => 100
      }
    }

    it 'mode' do
      output = LogStash::Outputs::Solr.new(config)
      output.register

      mode = output.instance_variable_get('@mode')
      expect(mode).to eq('SolrCloud')
    end
  end

  describe 'receive_standalone' do
    let(:config) {
      {
        'url' => 'http://localhost:8983/solr/collection1',
        'defined_fields' => ['id', 'title'],
        'ignore_undefined_fields' => true,
        'unique_key_field' => 'id',
        'timestamp_field' => 'event_timestamp',
        'flush_size' => 100
      }
    }

    let(:sample_record) {
      {
        'id' => 'change.me',
        'title' => 'change.me'
      }
    }

    it 'receive' do
      output = LogStash::Outputs::Solr.new(config)
      output.register

      output.receive(sample_record)
    end
  end

  describe 'receive_solrcloud' do
    before do
      start_zookeeper
    end

    after do
      stop_zookeeper
    end

    let(:config) {
      {
        'zk_host' => 'localhost:3292/solr',
        'collection' => 'collection1',
        'defined_fields' => ['id', 'title'],
        'ignore_undefined_fields' => true,
        'unique_key_field' => 'id',
        'timestamp_field' => 'event_timestamp',
        'flush_size' => 100
      }
    }

    let(:sample_record) {
      {
        'id' => 'change.me',
        'title' => 'change.me'
      }
    }

    it 'receive' do
      output = LogStash::Outputs::Solr.new(config)
      output.register

      output.receive(sample_record)
    end
  end

  def start_zookeeper
    @zk_server = ZK::Server.new do |config|
      config.client_port = 3292
      config.enable_jmx = true
      config.force_sync = false
    end

    @zk_server.run

    zk = ZK.new('localhost:3292')
    delete_nodes(zk, '/solr')
    create_nodes(zk, '/solr/live_nodes')
    create_nodes(zk, '/solr/collections')
    ['localhost:8983_solr'].each do |node|
      zk.create("/solr/live_nodes/#{node}", '', mode: :ephemeral)
    end
    ['collection1'].each do |collection|
      zk.create("/solr/collections/#{collection}")
      json = File.read("spec/files/collections/#{collection}/state.json")
      zk.create("/solr/collections/#{collection}/state.json", json, mode: :ephemeral)
    end
  end

  def stop_zookeeper
    @zk_server.shutdown
  end

  def delete_nodes(zk, path)
    zk.children(path).each do |node|
      delete_nodes(zk, File.join(path, node))
    end
    zk.delete(path)
  rescue ZK::Exceptions::NoNode
  end

  def create_nodes(zk, path)
    parent_path = File.dirname(path)
    unless zk.exists?(parent_path, :watch => true) then
      create_nodes(zk, parent_path)
    end
    zk.create(path)
  end
end




lib/logstash/outputs/solr.rb
# encoding: utf-8
require "logstash/outputs/base"
require "logstash/namespace"

require 'securerandom'
require "stud/buffer"
require 'rsolr'
require 'zk'
require 'rsolr/cloud'

# An Solr output that send data to Apache Solr.
class LogStash::Outputs::Solr < LogStash::Outputs::Base
  config_name "solr"

  include Stud::Buffer

  # The Solr server url (for example http://localhost:8983/solr/collection1).
  config :url, :validate => :string, :default => nil

  # The ZooKeeper connection string that SolrCloud refers to (for example localhost:2181/solr).
  config :zk_host, :validate => :string, :default => nil
  # The SolrCloud collection name.
  config :collection, :validate => :string, :default => 'collection1'

  # The defined fields in the Solr schema.xml. If omitted, it will get fields via Solr Schema API.
  config :defined_fields, :validate => :array, :default => nil
  # Ignore undefined fields in the Solr schema.xml.
  config :ignore_undefined_fields, :validate => :boolean, :default => false

  # A field name of unique key in the Solr schema.xml. If omitted, it will get unique key via Solr Schema API.
  config :unique_key_field, :validate => :string, :default => nil
  # A field name of event timestamp in the Solr schema.xml (default event_timestamp).
  config :timestamp_field, :validate => :string, :default => 'event_timestamp'

  # The batch size used in update.
  config :flush_size, :validate => :number, :default => 100

  # The batch size used in update.
  config :idle_flush_time, :validate => :number, :default => 10

  MODE_STANDALONE = 'Standalone'
  MODE_SOLRCLOUD = 'SolrCloud'

  public
  def register
    @mode = nil
    if ! @url.nil? then
      @mode = MODE_STANDALONE
    elsif ! @zk_host.nil?
      @mode = MODE_SOLRCLOUD
    end

    @solr = nil
    @zk = nil

    if @mode == MODE_STANDALONE then
      @solr = RSolr.connect :url => @url
    elsif @mode == MODE_SOLRCLOUD then
      @zk = ZK.new(@zk_host)
      cloud_connection = RSolr::Cloud::Connection.new(@zk)
      @solr = RSolr::Client.new(cloud_connection, read_timeout: 60, open_timeout: 60)
    end

    buffer_initialize(
      :max_items => @flush_size,
      :max_interval => @idle_flush_time,
      :logger => @logger
    )
  end # def register

  public
  def receive(event)
    buffer_receive(event)
  end # def event

  public
  def flush(events, close=false)
    documents = []

    @fields = @defined_fields.nil? || @defined_fields.empty? ? get_fields : @defined_fields

    @unique_key = @unique_key_field.nil? ? get_unique_key : @unique_key_field

    events.each do |event|
      document = event.to_hash()

      unless document.has_key?(@unique_key) then
        document.merge!({@unique_key => SecureRandom.uuid})
      end

      unless document.has_key?(@timestamp_field) then
        document.merge!({@timestamp_field => document['@timestamp']})
      end

      if @ignore_undefined_fields then
        document.each_key do |key|
          unless @fields.include?(key) then
            document.delete(key)
          end
        end
      end

      @logger.info 'Record: %s' % document.inspect

      documents.push(document)
    end

    if @mode == MODE_STANDALONE then
      @solr.add documents, :params => {:commit => true}
      @logger.info 'Added %d document(s) to Solr' % documents.count
    elsif @mode == MODE_SOLRCLOUD then
      @solr.add documents, collection: @collection, :params => {:commit => true}
      @logger.info 'Added %d document(s) to Solr' % documents.count
    end

    rescue Exception => e
      @logger.warn('An error occurred while indexing: #{e.message}')
  end # def flush

  public
  def close
    unless @zk.nil? then
      @zk.close
    end
  end # def close

  private
  def get_unique_key
    response = nil

    if @mode == MODE_STANDALONE then
      response = @solr.get 'schema/uniquekey'
    elsif @mode == MODE_SOLRCLOUD then
      response = @solr.get 'schema/uniquekey', collection: @collection
    end

    unique_key = response['uniqueKey']
    @logger.info 'Unique key: #{unique_key}'

    return unique_key

    rescue Exception => e
      @logger.warn 'Unique key: #{e.message}'
  end # def get_unique_key

  private
  def get_fields
    response = nil

    if @mode == MODE_STANDALONE then
      response = @solr.get 'schema/fields'
    elsif @mode == MODE_SOLRCLOUD then
      response = @solr.get 'schema/fields', collection: @collection
    end

    fields = []
    response['fields'].each do |field|
      fields.push(field['name'])
    end
    @logger.info 'Fields: #{fields}'

    return fields

    rescue Exception => e
      @logger.warn 'Fields: #{e.message}'
  end # def get_fields
end # class LogStash::Outputs::Solr
